drivers/network/netio/netio.c

```diff
@@ -0,0 +1,882 @@
+{
+	struct _WSK_SOCKET_INTERNAL *s = (struct _WSK_SOCKET_INTERNAL*) Socket;
+	NTSTATUS status = STATUS_NOT_IMPLEMENTED;
+
+	if (s == NULL) {
+		DbgPrint("WskControlSocket: Socket is NULL\n");
+		status = STATUS_INVALID_PARAMETER;
+	}
+	switch (RequestType) {
+	case WskSetOption:
+		switch (Level) {
+		case SOL_SOCKET:
+			switch (ControlCode) {
+			case SO_REUSEADDR:	/* add more supported flags here ... */
+				if (InputBuffer == NULL) {
+					DbgPrint("WskControlSocket: Need an InputBuffer for this operation\n");
+					status = STATUS_INVALID_PARAMETER;
+					break;
+				}
+				if (InputSize < 4) {
+					DbgPrint("WskControlSocket: InputBuffer too small for this operation\n");
+					status = STATUS_INVALID_PARAMETER;
+					break;
+				}
+				int flag = *(int*) InputBuffer;
+				if (flag != 0) {
+					s->Flags |= ControlCode;
+				} else {
+					s->Flags &= ~ControlCode;
+				}
+				status = STATUS_SUCCESS;
+
+				break;
+			/* Windows specific. This sets the mask for callback functions: */
+			case SO_WSK_EVENT_CALLBACK:
+				if (InputBuffer == NULL) {
+					DbgPrint("WskControlSocket: Need an InputBuffer for this operation\n");
+					status = STATUS_INVALID_PARAMETER;
+					break;
+				}
+				if (InputSize < sizeof(WSK_EVENT_CALLBACK_CONTROL)) {
+					DbgPrint("WskControlSocket: InputBuffer too small for this operation\n");
+					status = STATUS_INVALID_PARAMETER;
+					break;
+				}
+				WSK_EVENT_CALLBACK_CONTROL *c = (WSK_EVENT_CALLBACK_CONTROL*)InputBuffer;
+
+					/* TODO: compare this bytewise */
+/*				if (c->NpiId != &NPI_WSK_INTERFACE_ID) {
+					DbgPrint("WskControlSocket: Only NPI_WSK_INTERFACE_ID supported, sorry.\n");
+					status = STATUS_INVALID_PARAMETER;
+					break;
+				} */
+				s->CallbackMask = c->EventMask;
+					/* TODO: and start listening here? */
+
+				status = STATUS_SUCCESS;
+				break;
+
+			default:
+				DbgPrint("WskControlSocket: ControlCode %d Not implemented.\n", ControlCode);
+			}
+			break;
+		default:
+			DbgPrint("WskControlSocket: Level %d Not implemented.\n", Level);
+		}
+		break;
+
+	case WskGetOption:
+	case WskIoctl:
+	default:
+		DbgPrint("WskControlSocket: Option %d Not implemented.\n", RequestType);
+	}
+	if (Irp != NULL) {
+		Irp->IoStatus.Status = status;
+/* ??		Irp->IoStatus.Information = Irp->IoStatus.Information; */
+	}
+	return status;
+}
+
+static NTSTATUS WSKAPI WskCloseSocket(
+    _In_ PWSK_SOCKET Socket,
+    _Inout_ PIRP Irp)
+{
+	struct _WSK_SOCKET_INTERNAL *s = (struct _WSK_SOCKET_INTERNAL*) Socket;
+
+	SocketPut(s);
+	/* TODO: DummyCallDriver and IoComplete */
+	return STATUS_SUCCESS;
+}
+
+static struct _TRANSPORT_ADDRESS *TdiTransportAddressFromSocketAddress(PSOCKADDR SocketAddress)
+{
+	struct _TRANSPORT_ADDRESS *ta;
+
+	ta = ExAllocatePoolWithTag(NonPagedPool, sizeof(*ta)+sizeof(struct sockaddr) , 'ADDR');
+	if (ta == NULL) {
+		DbgPrint("TdiTransportAddressFromSocketAddress: Out of memory\n");
+		return NULL;
+	}
+
+//	memset(ta, 0, sizeof(*ta)+sizeof(struct sockaddr));
```

Comment:
when committing commented code that is intended to be added for the future, add some `// FIXME: to be added when condition xyz is met` or `// FIXME: find out why this doesn't work yet` or if it is a left-over: just strip it before committing.

Reply:
Ok removed that part